### PR TITLE
Implement svg output in notebooks #29

### PR DESCRIPTION
**Pull request by benabel**
This implementation use a simple algorithm in which the layout used is
based on a perfect tree of same height in which all leaves would be
regularly spaced.

This approach is ok for little trees to about 4 levels but for larger trees
with holes a better approach would be to implement the Reingold &
Tilford’s algorithm.

### CODE REVIEW

**Comment by benabel** [comment]
@joowani Currently the 2.7 and 3.5 tests fails because the code is using f-strings that were added in python 3.5.

https://docs.python.org/3/reference/lexical_analysis.html#f-strings

Do you to maintain compatibility with these python versions even if they have reached their end-of-life:

https://devguide.python.org/devcycle/#end-of-life-branches

**Comment by joowani** [comment]
Hi @benabel,

Thanks for the PR! Yes I am planning to drop support for those two, and I'm also planning to switch over to github actions etc as well. I'll keep this open and re-evaluate after.

**Comment by joowani** [comment]
Instead of implementing this method from scratch, I've decided to use [python-graphviz](https://pypi.org/project/graphviz) instead (it has many more features including Jupyter notebook support) in version 6.0.0. Please see the [documentation](https://binarytree.readthedocs.io/en/main/graphviz.html) for more details.

**Comment by benabel** [comment]
@joowani using phython-graphviz adds dependecies on python side and on system side with the graphviz binary. For users like teachers that don't have any rights on a machine this feature  will be unavailable. You could use a try except and fallback to my implementation if graphviz isn't available on the system.

File: binarytree/__init__.py [hunk 1]
```diff
@@ -16,6 +16,8 @@
     NodeReferenceError,
 )
 
+from binarytree.layout import _generate_svg
```
Comment (joowani) [inline binarytree/__init__.py:19]: Could you please rename module `layout` to `svg` and rename `_generate_svg` to `generate_svg`?

File: binarytree/__init__.py [hunk 2]
```diff
@@ -421,6 +423,23 @@ def __repr__(self):
         """
         return 'Node({})'.format(self.val)
 
+    def _repr_svg_(self):
+        """Return the svg representation on current node.
```
Comment (joowani) [inline binarytree/__init__.py:427]: This can be changed to `Return the svg representation of the binary tree`.

File: binarytree/__init__.py [hunk 3]
```diff
@@ -421,6 +423,23 @@ def __repr__(self):
...
+
+        :return: String representation.
+        :rtype: str | unicode
```
Comment (joowani) [inline binarytree/__init__.py:430]: This can just be `str` now.

File: binarytree/__init__.py [hunk 4]
```diff
@@ -421,6 +423,23 @@ def __repr__(self):
...
+
+        **Example**:
+
+        .. doctest::
+
+            >>> from binarytree import Node
+            >>>
+            >>> Node(1)
+            Node(1)
```
Comment (joowani) [inline binarytree/__init__.py:439]: This is not a meaningful doctest so it can be removed.

File: binarytree/layout.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
+""" Module containing layout related algorithms."""
+
+
+def _get_coords(values):
+    """Generate nodes and edges relative coordinates lists
```
Comment (joowani) [inline binarytree/layout.py:5]: Please change to `Generate the coordinates used for rendering the nodes and edges.`

File: binarytree/layout.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            edges.append((x//2, y-1, x, y))
+        x += 1
+        # check if level is full
+        if x == 2**y:
+            x = 0
+            y += 1
+    return nodes, edges
+
+
+def _generate_svg(values):
```
Comment (joowani) [inline binarytree/layout.py:40]: Rename to `generate_svg` as mentioned above.

File: tests/test_layout.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    assert _get_coords(values) == (
+        [(0, 0, 0), (0, 1, 6), (1, 1, 5), (1, 2, 1), (2, 2, 4), (3, 2, 2)],
+        [(0, 0, 0, 1), (0, 0, 1, 1), (0, 1, 1, 2), (1, 1, 2, 2), (1, 1, 3, 2)]
+        )
+
+
+def test_generate_svg():
+    svg = _generate_svg([0, 1, 2])
+    svg_tree = ET.fromstring(svg)
+    assert svg_tree.tag == "{http://www.w3.org/2000/svg}svg"
```
Comment (joowani) [inline tests/test_layout.py:19]: Please rename to `test_svg.py`

File: tests/test_layout.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
+from __future__ import absolute_import, unicode_literals
```
Comment (joowani) [inline tests/test_layout.py:1]: This can be removed.

File: examples/svg-output.ipynb [hunk 9]
```diff
@@ -0,0 +1,321 @@
+{
```
Comment (joowani) [inline examples/svg-output.ipynb:1]: I already have a gif so this file can be removed.

File: binarytree/layout.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
+""" Module containing layout related algorithms."""
+
+
+def _get_coords(values):
```
Comment (joowani) [inline binarytree/layout.py:4]: Since you are importing this outside of the module, should be renamed to `get_coords`.

**Review by joowani — CHANGES_REQUESTED** [review]
Awesome PR!

**Comment by joowani** [comment]
Hi @benabel,

I like your idea. I've reopened the PR and left some comments. Could you please address them and rebase? Thanks again for your contribution.

File: binarytree/layout.py (hunk 10 above)
Comment (benabel) [reply binarytree/layout.py:4]: This is only used on layout module and test.

**Comment by benabel** [comment]
@joowani I tried to amend the code using your suggestions, however I can't pass the doctests.

**Comment by benabel** [comment]
@joowani In fact the doctests fails due to the `graphviz` method. My changes seems ok, I added typings and unit tests passes.

2 Errors:
```
Failed example:
    graph.body              # Get the DOT body
Expected nothing
Got:
    ['\t140536390881360 [label="<l>|<v> 4|<r>"]', '\t140536390881360:l -> 140536390659728:v', '\t140536390881360:r -> 140536390660048:v', '\t140536390659728 [label="<l>|<v> 13|<r>"]', '\t140536390659728:l -> 140536390659472:v', '\t140536390659728:r -> 140536390657424:v', '\t140536390660048 [label="<l>|<v> 1|<r>"]', '\t140536390660048:l -> 140536390573584:...
```

```
graphviz.backend.ExecutableNotFound: failed to execute ['dot', '-Kdot', '-Tpdf', '-O', 'Digraph.gv'], make sure the Graphviz executables are on your systems' PATH
```

**Comment by joowani** [comment]
Hi @benabel,

No worries, I can take care of the clean up. Thanks again for your contribution! This is very useful.

**Comment by benabel** [comment]
You're welcome @joowani. Thanks for sharing this package.

**Pull request merged**